### PR TITLE
Add a fallback policy to be used whenever OPA server is unresponsive/unreachable

### LINKED ISSUE #218: [FEAT] Add some fallback functionality in the case of a failed connection to OPA
**Is your feature request related to a problem? Please describe.**
The issue is backstage will become crippled should the OPA server not be running, or is otherwise inaccessible. This is acceptable in some contexts, but it can be a critical problem in certain environments. In particular, it is a crucial issue blocking initial adoption of OPA in my organization, as concerns are raised about "what happens if OPA is down?".

**Describe the solution you'd like**
Some fallback or failsafe functionality. A simple option is a wildcard ALLOW policy in case OPA is inaccessible. Another option is falling back to a secondary permission policy implemented in Typescript. Ideally, this would be a configuration option to avoid causing undesired effects on existing projects that already implement the OPA Wrapper.

**Describe alternatives you've considered**
I've looked over through the code of he plugins, and plenty of Backstage documentation. There is no documented pathway to solve this without forking this plugin, as the Backstage-OPA plugins are exclusive with other permissions plugins (or an order of priority can't be established)

**Additional context**
I've been authorized to work on this issue by my company, and I'll try to have something to show by Saturday. I reached out to Parsifal-M on Mastodon and he said this would be an interesting feature to add. My implementation would seek to be of minimal impact to existing projects, and hopefully show an example of fallback policies.

### PR DESCRIPTION
**Pull request by coldblade2000**
## What type of PR is this?

Please check the category that applies:

- [X] Feature
- [ ] Bug Fix
- [ ] Refactor
- [ ] Documentation Update

## Description

Adds a failover mechanism to the wrapper plugin in the case of a failed connection with the OPA server.

### Additional changes

- Fixed some opaClient tests that weren't actually covering the intended lines they were meant to cover
- Added documentation for this change

## Related Issues

If applicable, please link any related issues. Use "Closes" keyword to automatically close the issue on PR merge.

Closes #218 

## Testing

- [X] I have included or updated unit tests for my changes.
- [ ] I would like some help writing tests.
- [ ] These changes do not require tests.

### CODE REVIEW

**Comment by Parsifal-M** [comment]
Wow fantastic work! 👏 

I'm out and about right now but can't wait to dig into this tomorrow!

Thanks so much for the work on this!

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts [hunk 1]
```diff
@@ -25,21 +26,27 @@ export class OpaClient {
...
+   * @param opaPolicyFallback - The fallback policy to use when the OPA server is unavailable or unresponsive. You can optionally provide the fallback policy here, otherwise it will be taken from the app-config.
    */
   async evaluatePolicy(
     input: PolicyEvaluationInput,
     opaEntryPoint?: string,
+    opaPolicyFallback?: string,
   ): Promise<PolicyEvaluationResult> {
     const setEntryPoint = opaEntryPoint ?? this.opaEntryPoint;
     const opaBaseUrl = this.opaBaseUrl;
+    const policyFallback = opaPolicyFallback ?? this.opaPolicyFallback;
```
Comment (Parsifal-M) [inline plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:49]: I think we can simplify this method a little bit, we don't need to pass the `opaPolicyFallback` to it and we can simply do the below, as its already being set during the instantiation of the class

```ts
const policyFallback = this.opaPolicyFallback;
```

WDYT?

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts [hunk 2]
```diff
@@ -88,12 +100,19 @@ export class OpaClient {
...
     } catch (error: unknown) {
-      this.logger.error(
-        `An error occurred while sending the policy input to the OPA server: ${error}`,
-      );
-      throw new Error(
-        `An error occurred while sending the policy input to the OPA server: ${error}`,
-      );
+      const message = `An error occurred while sending the policy input to the OPA server:`;
+
+      if (error instanceof Error && error.name === 'FetchError') {
```
Comment (Parsifal-M) [inline plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:105]: What do you think of doing this instead here?

```ts
import fetch, { FetchError } from 'node-fetch';
//...
if (error instanceof FetchError)
```

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts [hunk 3]
```diff
@@ -14,6 +14,7 @@ import { LoggerService } from '@backstage/backend-plugin-api';
 export class OpaClient {
   private readonly opaEntryPoint?: string;
   private readonly opaBaseUrl?: string;
+  private readonly opaPolicyFallback?: string;
```
Comment (Parsifal-M) [inline plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:17]: We _could_ create a type for this like the below in the `types.ts`

```ts
export type OpaFallbackPolicy = 'allow' | 'deny' | undefined;
```

```ts
private readonly opaPolicyFallback?: OpaFallbackPolicy;
```

WDYT?

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.test.ts
Comment (Parsifal-M) [inline plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.test.ts:1]: Just a thank you for making these more robust 😄 has been on my TODOs for quite some time!

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts [hunk 4]
```diff
@@ -73,12 +80,17 @@ export class OpaClient {
...
       if (!opaResponse.ok) {
-        this.logger.error(
-          `An error occurred while sending the policy input to the OPA server: ${opaResponse.status} - ${opaResponse.statusText}`,
-        );
-        throw new Error(
-          `An error occurred while sending the policy input to the OPA server: ${opaResponse.status} - ${opaResponse.statusText}`,
-        );
+        const message = `An error response was returned after sending the policy input to the OPA server: ${opaResponse.status} - ${opaResponse.statusText}`;
+
+        if (policyFallback?.toLowerCase() === 'allow') {
```
Comment (Parsifal-M) [inline plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:85]: Just a NIT, I think Backstage prefers using `.toLocaleLowerCase('en-US')`

**Review by Parsifal-M — COMMENTED** [review]
Hey! 👋 

I tested this locally with the scenarios and I love it! I think this is a really good addition and a really nice first contribution!

I've left a few comments to look over when ever you have the time, please let me know if you'd like some help getting this over the finish line I am happy to dive in 😄 

I will be semi-available tomorrow but from then on I will keep an eye on this! 👀 

Thanks again! 👏

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts (hunk 1 above)
Comment (coldblade2000) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:49]: I think that's a good idea, and I had thought about doing that. However, the thing is it makes it harder to test. Right now the Config has a mocked implementation, which makes it pretty hard to test different config values (as I have to alternate between 'allow' and 'deny'). I'm changing the tests a bit to make your change possible, LMK what you think. I'll link the commit once I've finished it

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts (hunk 2 above)
Comment (coldblade2000) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:105]: This line was the hardest to do. Essentially the mocking of `fetch` causes a lot of problems with actually testing a function with FetchError in it, as a mock node-fetch implementation will also make FetchError `undefined`, even with some `jest.requireActual` magic. That if statement is the most elegant solution I could come up with, surprisingly enough. To simplify it, fetch couldn't be mocked.

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts (hunk 3 above)
Comment (coldblade2000) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:17]: I thought of doing that but I didn't see any similar ones so I didn't do it. I'll make it happen then

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts (hunk 1 above)
Comment (coldblade2000) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts]: https://github.com/Parsifal-M/backstage-opa-plugins/pull/220/commits/3ba93512564d416f953f8b0584b02ff15e97c1d1

I dropped the ConfigReader.getOptionalString mock implementation, and replaced with just a base config object read by ConfigReader set in the beforeAll() of the jest tests. This makes it easier to test a slightly modified config.

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts (hunk 2 above)
Comment (coldblade2000) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:104]: I'd recommend adding a comment above explaining as such, but I prefer to ask you first.

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts (hunk 3 above)
Comment (coldblade2000) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts]: https://github.com/Parsifal-M/backstage-opa-plugins/pull/220/commits/76d83684eadf78d3e7a126608c2ba5f4731b5e63

LMK if this is okay.

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.test.ts
Comment (coldblade2000) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.test.ts:1]: Of course, glad to help!

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts (hunk 4 above)
Comment (coldblade2000) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts]: Done

**Comment by Parsifal-M** [comment]
Hey!

Kudos for all the hard work on this so far! Just letting you know I'll take a proper look tomorrow as I have some things to do this afternoon and won't get a chance 😢 

Thanks again!

File: plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts (hunk 2 above)
Comment (Parsifal-M) [reply plugins/permission-backend-module-opa-wrapper/src/opa-client/opaClient.ts:107]: Sure thats fine with me 😄

**Comment by Parsifal-M** [comment]
Hey @coldblade2000 👋 !

Thanks for all your hard work on this, lets get the actions to pass, think we just need to run a prettier, then I think this is okay to merge! :shipit: 

Appreciate all that you've done here so far!

**Comment by coldblade2000** [comment]
> Hey @coldblade2000 👋 !
> 
> Thanks for all your hard work on this, lets get the actions to pass, think we just need to run a prettier, then I think this is okay to merge! :shipit:
> 
> Appreciate all that you've done here so far!

Perfect. Yeah, I'd ran Lint and Prettier but my IDE was reformating my code on commit, so It was failing Prettier again. Fixed it

**Comment by Parsifal-M** [comment]
Thank you so much for this! Great idea 💡 and thanks for taking the time to implement it ☺️ 

I want to do some other tidying up that I've had on my to-do list for a while then I'll cut a release! 👏

**Pull request merged**